**You may need to install [OpenCV](https://pypi.python.org/pypi/opencv-python) and [scikit-video](http://www.scikit-video.org/stable/).**

In [0]:
! pip install scikit-video

    100% |████████████████████████████████| 2.3MB 12.0MB/s 


In [0]:
import keras
import numpy as np
import io
import base64
from IPython.display import HTML
import skvideo
import skvideo.io
import cv2
import json
skvideo.setFFmpegPath("/usr/local/bin")
import skvideo.io



from keras.models import Sequential,model_from_json
from keras.layers.core import Dense, Flatten
from keras.optimizers import sgd
from keras.layers import Conv2D, MaxPooling2D, Activation, AveragePooling2D,Reshape,BatchNormalization

/usr/local/lib/python3.6/dist-packages/skvideo/__init__.py:306: UserWarning: ffmpeg/ffprobe not found in path: /usr/local/bin
  warnings.warn("ffmpeg/ffprobe not found in path: " + str(path), UserWarning)


# MiniProject #3: Deep Reinforcement Learning

__Notations__: $E_p$ is the expectation under probability $p$. Please justify each of your answer and widely comment your code.

# Context

In a reinforcement learning algorithm, we modelize each step $t$ as an action $a_t$ obtained from a state $s_t$, i.e. $\{(a_{t},s_{t})_{t\leq T}\}$ having the Markov property. We consider a discount factor $\gamma \in [0,1]$ that ensures convergence. The goal is to find among all the policies $\pi$, one that maximizes the expected reward:

\begin{equation*}
R(\pi)=\sum_{t\leq T}E_{p^{\pi}}[\gamma^t r(s_{t},a_{t})] \> ,
\end{equation*}

where: 
\begin{equation*}p^{\pi}(a_{0},a_{1},s_{1},...,a_{T},s_{T})=p(a_{0})\prod_{t=1}^{T}\pi(a_{t}|s_{t})p(s_{t+1}|s_{t},a_{t}) \> .
\end{equation*}

We note the $Q$-function:

\begin{equation*}Q^\pi(s,a)=E_{p^{\pi}}[\sum_{t\leq T}\gamma^{t}r(s_{t},a_{t})|s_{0}=s,a_{0}=a] \> .
\end{equation*}

Thus, the optimal Q function is:
\begin{equation*}
Q^*(s,a)=\max_{\pi}Q^\pi(s,a) \> .
\end{equation*}

In this project, we will apply the deep reinforcement learning techniques to a simple game: an agent will have to learn from scratch a policy that will permit it maximizing a reward.

## The environment, the agent and the game

### The environment

```Environment``` is an abstract class that represents the states, rewards, and actions to obtain the new state.

In [0]:
class Environment(object):
    def __init__(self):
        pass

    def act(self, act):
        """
        One can act on the environment and obtain its reaction:
        - the new state
        - the reward of the new state
        - should we continue the game?

        :return: state, reward, game_over
        """
        pass


    def reset(self):
        """
        Reinitialize the environment to a random state and returns
        the original state

        :return: state
        """
        pass
    
    def draw(self):
        """
        Visualize in the console or graphically the current state
        """
        pass

The method ```act``` allows to act on the environment at a given state $s_t$ (stored internally), via action $a_t$. The method will return the new state $s_{t+1}$, the reward $r(s_{t},a_{t})$ and determines if $t\leq T$ (*game_over*).

The method ```reset``` simply reinitializes the environment to a random state $s_0$.

The method ```draw``` displays the current state $s_t$ (this is useful to check the behavior of the Agent).

We modelize $s_t$ as a tensor, while $a_t$ is an integer.

### The Agent

The goal of the ```Agent``` is to interact with the ```Environment``` by proposing actions $a_t$ obtained from a given state $s_t$ to attempt to maximize its __reward__ $r(s_t,a_t)$. We propose the following abstract class:

In [0]:
class Agent(object):
    def __init__(self, epsilon=0.1, n_action=4):
        self.epsilon = epsilon
        self.n_action = n_action
    
    def set_epsilon(self,e):
        self.epsilon = e

    def act(self,s,train=True):
        """ This function should return the next action to do:
        an integer between 0 and 4 (not included) with a random exploration of epsilon"""
        if train:
            if np.random.rand() <= self.epsilon:
                a = np.random.randint(0, self.n_action, size=1)[0]
            else:
                a = self.learned_act(s)
        else: # in some cases, this can improve the performance.. remove it if poor performances
            a = self.learned_act(s)

        return a

    def learned_act(self,s):
        """ Act via the policy of the agent, from a given state s
        it proposes an action a"""
        pass

    def reinforce(self, s, n_s, a, r, game_over_):
        """ This function is the core of the learning algorithm. 
        It takes as an input the current state s_, the next state n_s_
        the action a_ used to move from s_ to n_s_ and the reward r_.
        
        Its goal is to learn a policy.
        """
        pass

    def save(self):
        """ This function returns basic stats if applicable: the
        loss and/or the model"""
        pass

    def load(self):
        """ This function allows to restore a model"""
        pass



__Question 1__:
Explain the function act. Why is ```epsilon``` essential?

The function "act" separates two cases : training and testing. During training, the epsilon value intervenes for exploration. At each choice, the agent takes its action randomly with a probability epsilon. This is essential to allow the agent to explore other strategies that may have better outcomes than the ones it has been using before. 

After training, during the test phase, as the exploration part is over, we just want to exploit our discoveries, so the agent chooses the action with the best predicted outcome. 




***
### The Game

The ```Agent``` and the ```Environment``` work in an interlaced way as in the following (take some time to understand this code as it is the core of the project)

```python

epoch = 300
env = Environment()
agent = Agent()


# Number of won games
score = 0
loss = 0


for e in range(epoch):
    # At each epoch, we restart to a fresh game and get the initial state
    state = env.reset()
    # This assumes that the games will end
    game_over = False

    win = 0
    lose = 0
    
    while not game_over:
        # The agent performs an action
        action = agent.act(state)

        # Apply an action to the environment, get the next state, the reward
        # and if the games end
        prev_state = state
        state, reward, game_over = env.act(action)

        # Update the counters
        if reward > 0:
            win = win + reward
        if reward < 0:
            lose = lose -reward

        # Apply the reinforcement strategy
        loss = agent.reinforce(prev_state, state,  action, reward, game_over)

    # Save as a mp4
    if e % 10 == 0:
        env.draw(e)

    # Update stats
    score += win-lose

    print("Epoch {:03d}/{:03d} | Loss {:.4f} | Win/lose count {}/{} ({})"
          .format(e, epoch, loss, win, lose, win-lose))
    agent.save()
```

# The game, *eat cheese*

A rat runs on an island and tries to eat as much as possible. The island is subdivided into $N\times N$ cells, in which there are cheese (+0.5) and poisonous cells (-1). The rat has a visibility of 2 cells (thus it can see $5^2$ cells). The rat is given a time $T$ to accumulate as much food as possible. It can perform 4 actions: going up, down, left, right. 

The goal is to code an agent to solve this task that will learn by trial and error. We propose the following environment:

In [0]:
class Environment(object):
    def __init__(self, grid_size=10, max_time=500, temperature=0.1):
        grid_size = grid_size+4
        self.grid_size = grid_size
        self.max_time = max_time
        self.temperature = temperature

        #board on which one plays
        self.board = np.zeros((grid_size,grid_size))
        self.position = np.zeros((grid_size,grid_size))

        # coordinate of the cat
        self.x = 0
        self.y = 1

        # self time
        self.t = 0

        self.scale=16

        self.to_draw = np.zeros((max_time+2, grid_size*self.scale, grid_size*self.scale, 3))


    def draw(self,e):
        skvideo.io.vwrite(str(e) + '.mp4', self.to_draw)

    def get_frame(self,t):
        b = np.zeros((self.grid_size,self.grid_size,3))+128
        b[self.board>0,0] = 256
        b[self.board < 0, 2] = 256
        b[self.x,self.y,:]=256
        b[-2:,:,:]=0
        b[:,-2:,:]=0
        b[:2,:,:]=0
        b[:,:2,:]=0
        
        b =  cv2.resize(b, None, fx=self.scale, fy=self.scale, interpolation=cv2.INTER_NEAREST)

        self.to_draw[t,:,:,:]=b


    def act(self, action):
        """This function returns the new state, reward and decides if the
        game ends."""

        self.get_frame(int(self.t))

        self.position = np.zeros((self.grid_size, self.grid_size))

        self.position[0:2,:]= -1
        self.position[:,0:2] = -1
        self.position[-2:, :] = -1
        self.position[-2:, :] = -1

        self.position[self.x, self.y] = 1
        if action == 0:
            if self.x == self.grid_size-3:
                self.x = self.x-1
            else:
                self.x = self.x + 1
        elif action == 1:
            if self.x == 2:
                self.x = self.x+1
            else:
                self.x = self.x-1
        elif action == 2:
            if self.y == self.grid_size - 3:
                self.y = self.y - 1
            else:
                self.y = self.y + 1
        elif action == 3:
            if self.y == 2:
                self.y = self.y + 1
            else:
                self.y = self.y - 1
        else:
            RuntimeError('Error: action not recognized')

        self.t = self.t + 1
        reward = self.board[self.x, self.y]
        self.board[self.x, self.y] = 0
        game_over = self.t > self.max_time
        state = np.concatenate((self.board.reshape(self.grid_size, self.grid_size,1),
                        self.position.reshape(self.grid_size, self.grid_size,1)),axis=2)
        state = state[self.x-2:self.x+3,self.y-2:self.y+3,:]

        return state, reward, game_over

    def reset(self):
        """This function resets the game and returns the initial state"""

        self.x = np.random.randint(3, self.grid_size-3, size=1)[0]
        self.y = np.random.randint(3, self.grid_size-3, size=1)[0]


        bonus = 0.5*np.random.binomial(1,self.temperature,size=self.grid_size**2)
        bonus = bonus.reshape(self.grid_size,self.grid_size)

        malus = -1.0*np.random.binomial(1,self.temperature,size=self.grid_size**2)
        malus = malus.reshape(self.grid_size, self.grid_size)

        self.to_draw = np.zeros((self.max_time+2, self.grid_size*self.scale, self.grid_size*self.scale, 3))


        malus[bonus>0]=0

        self.board = bonus + malus

        self.position = np.zeros((self.grid_size, self.grid_size))
        self.position[0:2,:]= -1
        self.position[:,0:2] = -1
        self.position[-2:, :] = -1
        self.position[-2:, :] = -1
        self.board[self.x,self.y] = 0
        self.t = 0

        state = np.concatenate((
                               self.board.reshape(self.grid_size, self.grid_size,1),
                        self.position.reshape(self.grid_size, self.grid_size,1)),axis=2)

        state = state[self.x - 2:self.x + 3, self.y - 2:self.y + 3, :]
        return state

The following elements are important because they correspond to the hyper parameters for this project:

In [0]:
# parameters
size = 13
T=200
temperature=0.3
epochs_train=50 # set small when debugging
epochs_test=10 # set small when debugging

# display videos
def display_videos(name):
    video = io.open(name, 'r+b').read()
    encoded = base64.b64encode(video)
    return '''<video alt="test" controls>
                <source src="data:video/mp4;base64,{0}" type="video/mp4" />
             </video>'''.format(encoded.decode('ascii'))

__Question 2__ Explain the use of the arrays ```position``` and ```board```.

The Board stocks the information concerning the content of each cell ( cheese, poisons) and therefore is important for attributing the reward after each action.  \\

Position shows which cells are visible by the agent, and on which cell is it. It also gives information about the boundaries in the map.

## Random Agent

***
__Question 3__ Implement a random Agent (only ```learned_act``` needs to be implemented):

In [0]:
class RandomAgent(Agent):
    def __init__(self):
        super(RandomAgent, self).__init__()
        pass

    def learned_act(self, s):
        return np.random.randint(4)
        

***
***
__Question 4__ Visualize the game moves. You need to fill in the following function for the evaluation:

In [0]:
def test(agent,env,epochs,prefix=''):
    # Number of won games
    score = 0
        
    for e in range(epochs):
      state = env.reset()
      # This assumes that the games will end
      game_over = False

      win = 0
      lose = 0

      while not game_over:
          # The agent performs an action
          action = agent.act(state)

          # Apply an action to the environment, get the next state, the reward
          # and if the games end
          prev_state = state
          state, reward, game_over = env.act(action)

          # Update the counters
          if reward > 0:
              win = win + reward
          if reward < 0:
              lose = lose -reward

          # Apply the reinforcement strategy
          loss = agent.reinforce(prev_state, state,  action, reward, game_over)

        
        
        ##### FILL IN HERE
        
      # Save as a mp4
      env.draw(prefix+str(e))

      # Update stats
      score = score + win-lose

      print("Win/lose count {}/{}. Average score ({})"
            .format(win, lose, score/(1+e)))
      print('Final score: '+str(score/epochs))

In [0]:
# Initialize the game
env = Environment(grid_size=size, max_time=T,temperature=temperature)

# Initialize the agent!
agent = RandomAgent()

test(agent,env,epochs_test,prefix='random')
HTML(display_videos('random0.mp4'))

Win/lose count 14.0/18.0. Average score (-4.0)
Final score: -0.4
Win/lose count 10.0/12.0. Average score (-3.0)
Final score: -0.6
Win/lose count 9.5/17.0. Average score (-4.5)
Final score: -1.35
Win/lose count 12.0/11.0. Average score (-3.125)
Final score: -1.25
Win/lose count 12.0/12.0. Average score (-2.5)
Final score: -1.25
Win/lose count 11.0/11.0. Average score (-2.0833333333333335)
Final score: -1.25
Win/lose count 12.0/6.0. Average score (-0.9285714285714286)
Final score: -0.65
Win/lose count 5.0/7.0. Average score (-1.0625)
Final score: -0.85
Win/lose count 11.0/20.0. Average score (-1.9444444444444444)
Final score: -1.75
Win/lose count 9.0/9.0. Average score (-1.75)
Final score: -1.75


***
## DQN

Let us assume here that $T=\infty$.

***
__Question 5__ Let $\pi$ be a policy, show that:

\begin{equation*}
Q^{\pi}(s,a)=E_{(s',a')\sim p(.|s,a)}[r(s,a)+\gamma Q^{\pi}(s',a')]
\end{equation*}

Then, show that for the optimal policy $\pi^*$ (we assume its existence), the following holds: 

\begin{equation*}
Q^{*}(s,a)=E_{s'\sim \pi^*(.|s,a)}[r(s,a)+\gamma\max_{a'}Q^{*}(s',a')].
\end{equation*}
Finally, deduce that a plausible objective is:

\begin{equation*}
\mathcal{L}(\theta)=E_{s' \sim \pi^*(.|s,a)}\Vert r+\gamma\max\max_{a'}Q(s',a',\theta)-Q(s,a,\theta)\Vert^{2}.
\end{equation*}







If our Q-table depicts the optimal policy, the previous equality holds. Therefore, the absolute difference between the two parts of the equation is a suitable loss objective for our problem. This is also the case for the squared difference. 
and  : 
\begin{equation*}
\mathcal{L}(\theta)=E_{s' \sim \pi^*(.|s,a)}\Vert r+\gamma\max\max_{a'}Q(s',a',\theta)-Q(s,a,\theta)\Vert^{2}.
\end{equation*}

is a plausible objective for our learning problem.

***
The DQN-learning algorithm relies on these derivations to train the parameters $\theta$ of a Deep Neural Network:

1. At the state $s_t$, select the action $a_t$ with best reward using $Q_t$ and store the results;

2. Obtain the new state $s_{t+1}$ from the environment $p$;

3. Store $(s_t,a_t,s_{t+1})$;

4. Obtain $Q_{t+1}$ by minimizing  $\mathcal{L}$ from a recovered batch from the previously stored results.

***
__Question 6__ Implement the class ```Memory``` that stores moves (in a replay buffer) via ```remember``` and provides a ```random_access``` to these. Specify a maximum memory size to avoid side effects. You can for example use a ```list()``` and set by default ```max_memory=100```.

In [0]:
class Memory(object):
    def __init__(self, max_memory=100):
        self.max_memory = max_memory
        self.memory = list()

    def remember(self, m):
        self.memory.append(m)
        if len(self.memory)>self.max_memory:
            del self.memory[0]
            # Pop the first element if the memory is exceeded

    def random_access(self):
      return (self.memory[np.random.randint(len(self.memory))])


***
The pipeline we will use for training is given below:

In [0]:
def train(agent,env,epoch,prefix=''):
    # Number of won games
    score = 0
    loss = 0

    for e in range(epoch):
        print("new episode : " + str(e))
        # At each epoch, we restart to a fresh game and get the initial state
        state = env.reset()
        # This assumes that the games will terminate
        game_over = False

        win = 0
        lose = 0

        while not game_over:
            # The agent performs an action
            
            action = agent.act(state)

            # Apply an action to the environment, get the next state, the reward
            # and if the games end
            prev_state = state
            state, reward, game_over = env.act(action)

            # Update the counters
            if reward > 0:
                win = win + reward
            if reward < 0:
                lose = lose -reward

            # Apply the reinforcement strategy
            loss = agent.reinforce(prev_state, state,  action, reward, game_over)

        # Save as a mp4
        if e % 10 == 0:
            env.draw(prefix+str(e))

        # Update stats
        score += win-lose

        print("Epoch {:03d}/{:03d} | Loss {:.4f} | Win/lose count {}/{} ({})"
              .format(e, epoch, loss, win, lose, win-lose))
        agent.save(name_weights=prefix+'model.h5',name_model=prefix+'model.json')

***
__Question 7__ Implement the DQN training algorithm using a cascade of fully connected layers. You can use different learning rate, batch size or memory size parameters. In particular, the loss might oscillate while the player will start to win the games. You have to find a good criterium.

In [0]:
class DQN(Agent):
    def __init__(self, grid_size,  epsilon = 0.1, memory_size=100, batch_size = 16,n_state=2):
        super(DQN, self).__init__(epsilon = epsilon)

        # Discount for Q learning
        self.discount = 0.99
        
        self.grid_size = grid_size
        
        # number of state
        self.n_state = n_state

        # Memory
        
        self.memory = Memory(memory_size)
        
        # Batch size when learning
        self.batch_size = batch_size

    def learned_act(self, s):
        predictions =  self.model.predict(np.reshape(s, (1,5,5,self.n_state)))
        action = np.argmax(predictions[0,:])
        return action
    def reinforce(self, s_, n_s_, a_, r_, game_over_):
        # Two steps: first memorize the states, second learn from the pool
        self.memory.remember([s_, n_s_, a_, r_, game_over_])
        
        input_states = np.zeros((self.batch_size, 5,5,self.n_state))
        target_q = np.zeros((self.batch_size, 4))
        
        for i in range(self.batch_size):
            s_, n_s_, a_, r_, game_over_ =  self.memory.random_access()
            input_states[i] = s_
            q_target = self.model.predict(np.reshape(s_, (1,5,5,self.n_state)))
            target_q [i] = q_target

            if game_over_:
              target_q[i,a_] = r_
            else:
              target_q[i, a_] = r_  + self.discount *np.max(self.model.predict(np.reshape(n_s_, (1,5,5,self.n_state))))
        # HINT: Clip the target to avoid exploiding gradients.. -- clipping is a bit tighter
        
        target_q = np.clip(target_q, -3, 3)

        l = self.model.train_on_batch(input_states, target_q)


        return l

    def save(self,name_weights='model.h5',name_model='model.json'):
        self.model.save_weights(name_weights, overwrite=True)
        with open(name_model, "w") as outfile:
            json.dump(self.model.to_json(), outfile)
            
    def load(self,name_weights='model.h5',name_model='model.json'):
        with open(name_model, "r") as jfile:
            model = model_from_json(json.load(jfile))
        model.load_weights(name_weights)
        model.compile("sgd", "mse")
        self.model = model

            
class DQN_FC(DQN):
    def __init__(self, *args, lr=0.1,**kwargs):
        super(DQN_FC, self).__init__( *args,**kwargs)
        
        # NN Model
        model = Sequential()
        model.add(Flatten(input_shape=((5,5,2))))
        model.add(Dense(64, activation = "relu"))
        # 4 = Output Shape , number of possible actions  
            
        model.add(Dense(4))
        
        model.compile(sgd(lr=lr, decay=1e-4, momentum=0.0), "mse")
        model.summary()
        self.model = model
        

In [0]:
env = Environment(grid_size=size, max_time=T, temperature=0.3)
agent = DQN_FC(size, lr=.1, epsilon = 0.1, memory_size=2000, batch_size = 32)
train(agent, env, epochs_train, prefix='fc_train')
HTML(display_videos('fc_train0.mp4'))

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_3 (Flatten)          (None, 50)                0         
_________________________________________________________________
dense_5 (Dense)              (None, 64)                3264      
_________________________________________________________________
dense_6 (Dense)              (None, 4)                 260       
Total params: 3,524
Trainable params: 3,524
Non-trainable params: 0
_________________________________________________________________
new episode : 0
Epoch 000/050 | Loss 0.0068 | Win/lose count 5.5/8.0 (-2.5)
new episode : 1
Epoch 001/050 | Loss 0.0111 | Win/lose count 2.0/1.0 (1.0)
new episode : 2
Epoch 002/050 | Loss 0.0128 | Win/lose count 1.0/8.0 (-7.0)
new episode : 3
Epoch 003/050 | Loss 0.0190 | Win/lose count 5.5/5.0 (0.5)
new episode : 4
Epoch 004/050 | Loss 0.0083 | Win/lose count 3.0/3.0 (0.0)
new episode : 5
Epoch 005/0

***
***
__Question 8__ Implement the DQN training algorithm using a CNN (for example, 2 convolutional layers and one final fully connected layer).

In [0]:
class DQN_CNN(DQN):
    def __init__(self, *args,lr=0.1,**kwargs):
        super(DQN_CNN, self).__init__(*args,**kwargs)
        model = Sequential()
        model.add(Conv2D(50,(2,2),input_shape=(5,5,self.n_state,),activation='relu'))
        model.add(Conv2D(30,(2,2),activation='relu'))
        model.add(Flatten())
        model.add(Dense(4))
        model.compile(sgd(lr=lr, decay=1e-4, momentum=0.0), "mse")
        self.model = model


In [0]:
env = Environment(grid_size=size, max_time=T, temperature=0.3)
agent = DQN_CNN(size, lr=.1, epsilon = 0.1, memory_size=2000, batch_size = 32)
train(agent,env,epochs_train,prefix='cnn_train')
HTML(display_videos('cnn_train0.mp4'))

new episode : 0
Epoch 000/050 | Loss 0.0218 | Win/lose count 3.5/2.0 (1.5)
new episode : 1
Epoch 001/050 | Loss 0.0045 | Win/lose count 1.0/5.0 (-4.0)
new episode : 2
Epoch 002/050 | Loss 0.0048 | Win/lose count 1.0/4.0 (-3.0)
new episode : 3
Epoch 003/050 | Loss 0.0030 | Win/lose count 4.5/2.0 (2.5)
new episode : 4
Epoch 004/050 | Loss 0.0032 | Win/lose count 6.0/4.0 (2.0)
new episode : 5
Epoch 005/050 | Loss 0.0017 | Win/lose count 3.0/1.0 (2.0)
new episode : 6
Epoch 006/050 | Loss 0.0035 | Win/lose count 4.5/5.0 (-0.5)
new episode : 7
Epoch 007/050 | Loss 0.0013 | Win/lose count 4.0/3.0 (1.0)
new episode : 8
Epoch 008/050 | Loss 0.0464 | Win/lose count 6.0/4.0 (2.0)
new episode : 9
Epoch 009/050 | Loss 0.0021 | Win/lose count 7.5/1.0 (6.5)
new episode : 10
Epoch 010/050 | Loss 0.0026 | Win/lose count 7.0/3.0 (4.0)
new episode : 11
Epoch 011/050 | Loss 0.0012 | Win/lose count 9.5/3.0 (6.5)
new episode : 12
Epoch 012/050 | Loss 0.0039 | Win/lose count 7.5/3.0 (4.5)
new episode : 13
Ep

***
***
__Question 9__ Test both algorithms and compare their performances. Which issue(s) do you observe? Observe also different behaviors by changing the temperature.

In [0]:
# Temperature here controls the probability for a cell to contain a bonus or a malus. The higher the temperature, more often bonuses and maluses will be encountered. 

# We will start by temperature = 0.2

env = Environment(grid_size=size, max_time=T,temperature=0.2)
agent_cnn = DQN_CNN(size, lr=.1, epsilon = 0.1, memory_size=2000, batch_size = 32)
agent_cnn.load(name_weights='cnn_trainmodel.h5',name_model='cnn_trainmodel.json')

agent_fc = DQN_FC(size, lr=.1, epsilon = 0.1, memory_size=2000, batch_size = 32)
agent_cnn.load(name_weights='fc_trainmodel.h5',name_model='fc_trainmodel.json')
print('Test of the CNN')
test(agent_cnn,env,epochs_test,prefix='cnn_test')
print('Test of the FC')
test(agent_fc,env,epochs_test,prefix='fc_test')

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_10 (Flatten)         (None, 50)                0         
_________________________________________________________________
dense_15 (Dense)             (None, 64)                3264      
_________________________________________________________________
dense_16 (Dense)             (None, 4)                 260       
Total params: 3,524
Trainable params: 3,524
Non-trainable params: 0
_________________________________________________________________
Test of the CNN
Win/lose count 7.0/4.0. Average score (3.0)
Final score: 0.3
Win/lose count 3.0/1.0. Average score (2.5)
Final score: 0.5
Win/lose count 6.0/0. Average score (3.6666666666666665)
Final score: 1.1
Win/lose count 9.5/1.0. Average score (4.875)
Final score: 1.95
Win/lose count 1.5/1.0. Average score (4.0)
Final score: 2.0
Win/lose count 4.5/0. Average score (4.083333333333333)
Final score

In [0]:
HTML(display_videos('cnn_test0.mp4'))

In [0]:
HTML(display_videos('fc_test0.mp4'))

In [0]:
#Temperature = 0.5 

env = Environment(grid_size=size, max_time=T,temperature=0.5)
agent_cnn = DQN_CNN(size, lr=.1, epsilon = 0.1, memory_size=2000, batch_size = 32)
agent_cnn.load(name_weights='cnn_trainmodel.h5',name_model='cnn_trainmodel.json')

agent_fc = DQN_FC(size, lr=.1, epsilon = 0.1, memory_size=2000, batch_size = 32)
agent_cnn.load(name_weights='fc_trainmodel.h5',name_model='fc_trainmodel.json')
print('Test of the CNN')
test(agent_cnn,env,epochs_test,prefix='cnn_test')
print('Test of the FC')
test(agent_fc,env,epochs_test,prefix='fc_test')

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_14 (Flatten)         (None, 50)                0         
_________________________________________________________________
dense_20 (Dense)             (None, 64)                3264      
_________________________________________________________________
dense_21 (Dense)             (None, 4)                 260       
Total params: 3,524
Trainable params: 3,524
Non-trainable params: 0
_________________________________________________________________
Test of the CNN
Win/lose count 4.0/6.0. Average score (-2.0)
Final score: -0.2
Win/lose count 3.5/1.0. Average score (0.25)
Final score: 0.05
Win/lose count 2.5/4.0. Average score (-0.3333333333333333)
Final score: -0.1
Win/lose count 15.5/2.0. Average score (3.125)
Final score: 1.25
Win/lose count 13.0/4.0. Average score (4.3)
Final score: 2.15
Win/lose count 15.5/1.0. Average score (6.0)
Final score

In [0]:
# And now with temperature  = 0.8

env = Environment(grid_size=size, max_time=T,temperature=0.8)
agent_cnn = DQN_CNN(size, lr=.1, epsilon = 0.1, memory_size=2000, batch_size = 32)
agent_cnn.load(name_weights='cnn_trainmodel.h5',name_model='cnn_trainmodel.json')

agent_fc = DQN_FC(size, lr=.1, epsilon = 0.1, memory_size=2000, batch_size = 32)
agent_cnn.load(name_weights='fc_trainmodel.h5',name_model='fc_trainmodel.json')
print('Test of the CNN')
test(agent_cnn,env,epochs_test,prefix='cnn_test')
print('Test of the FC')
test(agent_fc,env,epochs_test,prefix='fc_test')

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_8 (Flatten)          (None, 50)                0         
_________________________________________________________________
dense_12 (Dense)             (None, 64)                3264      
_________________________________________________________________
dense_13 (Dense)             (None, 4)                 260       
Total params: 3,524
Trainable params: 3,524
Non-trainable params: 0
_________________________________________________________________
Test of the CNN
Win/lose count 36.0/2.0. Average score (34.0)
Final score: 3.4
Win/lose count 48.0/3.0. Average score (39.5)
Final score: 7.9
Win/lose count 35.5/1.0. Average score (37.833333333333336)
Final score: 11.35
Win/lose count 32.0/4.0. Average score (35.375)
Final score: 14.15
Win/lose count 37.5/5.0. Average score (34.8)
Final score: 17.4
Win/lose count 35.0/4.0. Average score (34.166666666

In [0]:
HTML(display_videos('cnn_test0.mp4'))

In [0]:
HTML(display_videos('fc_test0.mp4'))

###Observations 

We can first observe the effect of the temperature. Increasing the temperature highly increases the average score. Even if the maluses are also increased by temperature, our agents have learned to avoid them, and therefore mainly take advantage of the higher bonuses. For instance, the CNN average score goes from 5 with temperature = 0.2 to 31.2 with temperature = 0.8.
The FC goes from -1.35 to 5.95 average score.

The CNN architecture clearly outperforms the FC one. The results above show the gap in the average score between the two architectures. We can see in the video, that the CNN is more keen to explore the map, and avoids better the maluses. On the medium temperature example, T= 0.5, CNN reaches an average score of 9.9, while FC reaches 2.6.

Finally, exploration is still a week point. When we watch the videos, rarely more than half of the map is explored in each case. 



One main issue is observed .



***

The algorithm tends to not explore the map which can be an issue. We propose two ideas in order to encourage exploration:
1. Incorporating a decreasing $\epsilon$-greedy exploration. You can use the method ```set_epsilon```
2. Append via the environment a new state that describes if a cell has been visited or not

***
__Question 10__ Design a new ```train_explore``` function and environment class ```EnvironmentExploring``` to tackle the issue of exploration.



In [0]:
def train_explore(agent,env,epoch,prefix='explore'):
    # Number of won games
    score = 0
    loss = 0
    #Initialization of decreasing epsilon greedy exploration
    agent.set_epsilon(0.5) 
    discount_on_epsilon = 0.75
    for e in range(epoch):
        # Decreasing epsilon 
        agent.set_epsilon(agent.epsilon*discount_on_epsilon)
        print("new episode : " + str(e))
        # At each epoch, we restart to a fresh game and get the initial state
        state = env.reset()
        # This assumes that the games will terminate
        game_over = False

        win = 0
        lose = 0

        while not game_over:
            # The agent performs an action
            
            action = agent.act(state)

            # Apply an action to the environment, get the next state, the reward
            # and if the games end
            prev_state = state
            state, reward, game_over = env.act(action, train=True)

            # Update the counters
            if reward > 0:
                win = win + reward
            if reward < 0:
                lose = lose -reward

            # Apply the reinforcement strategy
            loss = agent.reinforce(prev_state, state,  action, reward, game_over)

        # Save as a mp4
        if e % 10 == 0:
            env.draw(prefix+str(e))

        # Update stats
        score += win-lose

        print("Epoch {:03d}/{:03d} | Loss {:.4f} | Win/lose count {}/{} ({})"
              .format(e, epoch, loss, win, lose, win-lose))
        agent.save(name_weights=prefix+'model.h5',name_model=prefix+'model.json')
        
class EnvironmentExploring(object):
    def __init__(self, grid_size=10, max_time=500, temperature=0.1):
        grid_size = grid_size+4
        self.grid_size = grid_size
        self.max_time = max_time
        self.temperature = temperature

        #board on which one plays
        self.board = np.zeros((grid_size,grid_size))
        self.position = np.zeros((grid_size,grid_size))
        self.malus_position = np.zeros((grid_size, grid_size)) # Motivating exploration

        # coordinate of the cat
        self.x = 0
        self.y = 1

        # self time
        self.t = 0

        self.scale=16

        self.to_draw = np.zeros((max_time+2, grid_size*self.scale, grid_size*self.scale, 3))


    def draw(self,e):
        skvideo.io.vwrite(str(e) + '.mp4', self.to_draw)

    def get_frame(self,t):
        b = np.zeros((self.grid_size,self.grid_size,3))+128
        b[self.board>0,0] = 256
        b[self.board < 0, 2] = 256
        b[self.x,self.y,:]=256
        b[-2:,:,:]=0
        b[:,-2:,:]=0
        b[:2,:,:]=0
        b[:,:2,:]=0
        
        b =  cv2.resize(b, None, fx=self.scale, fy=self.scale, interpolation=cv2.INTER_NEAREST)

        self.to_draw[t,:,:,:]=b


    def act(self, action, train):
        """This function returns the new state, reward and decides if the
        game ends."""

        self.get_frame(int(self.t))

        self.position = np.zeros((self.grid_size, self.grid_size))

        self.position[0:2,:]= -1
        self.position[:,0:2] = -1
        self.position[-2:, :] = -1
        self.position[-2:, :] = -1

        self.position[self.x, self.y] = 1
        if action == 0:
            if self.x == self.grid_size-3:
                self.x = self.x-1
            else:
                self.x = self.x + 1
        elif action == 1:
            if self.x == 2:
                self.x = self.x+1
            else:
                self.x = self.x-1
        elif action == 2:
            if self.y == self.grid_size - 3:
                self.y = self.y - 1
            else:
                self.y = self.y + 1
        elif action == 3:
            if self.y == 2:
                self.y = self.y + 1
            else:
                self.y = self.y - 1
        else:
            RuntimeError('Error: action not recognized')

        self.t = self.t + 1
        reward = 0 
        if train:
          reward = -self.malus_position[self.x, self.y]
        #When visited we update malus position
        self.malus_position[self.x, self.y] = 0.1

        reward = self.board[self.x, self.y]
        self.board[self.x, self.y] = 0
        game_over = self.t > self.max_time
        #We add the malus_position grid
        state = np.concatenate((self.malus_position.reshape(self.grid_size, self.grid_size,1),
                                self.board.reshape(self.grid_size, self.grid_size,1),
                        self.position.reshape(self.grid_size, self.grid_size,1)),axis=2)
        state = state[self.x-2:self.x+3,self.y-2:self.y+3,:]

        return state, reward, game_over

    def reset(self):
        """This function resets the game and returns the initial state"""

        self.x = np.random.randint(3, self.grid_size-3, size=1)[0]
        self.y = np.random.randint(3, self.grid_size-3, size=1)[0]


        bonus = 0.5*np.random.binomial(1,self.temperature,size=self.grid_size**2)
        bonus = bonus.reshape(self.grid_size,self.grid_size)

        malus = -1.0*np.random.binomial(1,self.temperature,size=self.grid_size**2)
        malus = malus.reshape(self.grid_size, self.grid_size)

        self.to_draw = np.zeros((self.max_time+2, self.grid_size*self.scale, self.grid_size*self.scale, 3))


        malus[bonus>0]=0

        self.board = bonus + malus

        self.position = np.zeros((self.grid_size, self.grid_size))
        self.position[0:2,:]= -1
        self.position[:,0:2] = -1
        self.position[-2:, :] = -1
        self.position[-2:, :] = -1
        self.board[self.x,self.y] = 0
        self.t = 0
        #Malus position is initialized to zero
        self.malus_position = np.zeros((self.grid_size, self.grid_size))
        
        
        state = np.concatenate((self.malus_position.reshape(self.grid_size, self.grid_size,1),
                                self.board.reshape(self.grid_size, self.grid_size,1),
                        self.position.reshape(self.grid_size, self.grid_size,1)),axis=2)

        state = state[self.x - 2:self.x + 3, self.y - 2:self.y + 3, :]
        return state
"""
## use those samples of code:
#In train explore:
state, reward, game_over = env.act(action, train=True)

## In Environment exploring:
# You will have to change n_state to 3 because you will use one more layer!
reward = 0
if train:
    reward = -self.malus_position[self.x, self.y]
self.malus_position[self.x, self.y] = 0.1

reward = reward + self.board[self.x, self.y]
# 3 "feature" states instead of 2
state = np.concatenate((self.malus_position.reshape(self.grid_size, self.grid_size,1),
                                self.board.reshape(self.grid_size, self.grid_size,1),
                        self.position.reshape(self.grid_size, self.grid_size,1)),axis=2)
                        
"""

'\n## use those samples of code:\n#In train explore:\nstate, reward, game_over = env.act(action, train=True)\n\n## In Environment exploring:\n# You will have to change n_state to 3 because you will use one more layer!\nreward = 0\nif train:\n    reward = -self.malus_position[self.x, self.y]\nself.malus_position[self.x, self.y] = 0.1\n\nreward = reward + self.board[self.x, self.y]\n# 3 "feature" states instead of 2\nstate = np.concatenate((self.malus_position.reshape(self.grid_size, self.grid_size,1),\n                                self.board.reshape(self.grid_size, self.grid_size,1),\n                        self.position.reshape(self.grid_size, self.grid_size,1)),axis=2)\n                        \n'

In [0]:
# Training
env = EnvironmentExploring(grid_size=size, max_time=T, temperature=0.5)
agent = DQN_CNN(size, lr=.1, epsilon = 0.5, memory_size=2000, batch_size = 32,n_state=3)
train_explore(agent, env, epochs_train, prefix='cnn_train_explore')
HTML(display_videos('cnn_train_explore0.mp4'))

new episode : 0
Epoch 000/050 | Loss 0.0074 | Win/lose count 9.0/8.0 (1.0)
new episode : 1
Epoch 001/050 | Loss 0.0070 | Win/lose count 3.0/11.0 (-8.0)
new episode : 2
Epoch 002/050 | Loss 0.0089 | Win/lose count 13.5/5.0 (8.5)
new episode : 3
Epoch 003/050 | Loss 0.0730 | Win/lose count 10.5/5.0 (5.5)
new episode : 4
Epoch 004/050 | Loss 0.0030 | Win/lose count 4.0/3.0 (1.0)
new episode : 5
Epoch 005/050 | Loss 0.0041 | Win/lose count 9.0/4.0 (5.0)
new episode : 6
Epoch 006/050 | Loss 0.0014 | Win/lose count 3.0/1.0 (2.0)
new episode : 7
Epoch 007/050 | Loss 0.0025 | Win/lose count 15.0/1.0 (14.0)
new episode : 8
Epoch 008/050 | Loss 0.0010 | Win/lose count 2.0/2.0 (0.0)
new episode : 9
Epoch 009/050 | Loss 0.0036 | Win/lose count 16.0/1.0 (15.0)
new episode : 10
Epoch 010/050 | Loss 0.0029 | Win/lose count 4.0/1.0 (3.0)
new episode : 11
Epoch 011/050 | Loss 0.0018 | Win/lose count 19.5/1.0 (18.5)
new episode : 12
Epoch 012/050 | Loss 0.0035 | Win/lose count 27.5/0 (27.5)
new episode 

In [0]:
def testexp(agent,env,epochs,prefix=''):
    # Number of won games
    score = 0
        
    for e in range(epochs):
      state = env.reset()
      # This assumes that the games will end
      game_over = False

      win = 0
      lose = 0

      while not game_over:
          # The agent performs an action
          action = agent.act(state)

          # Apply an action to the environment, get the next state, the reward
          # and if the games end
          prev_state = state
          state, reward, game_over = env.act(action, train=False)

          # Update the counters
          if reward > 0:
              win = win + reward
          if reward < 0:
              lose = lose -reward

          # Apply the reinforcement strategy
          loss = agent.reinforce(prev_state, state,  action, reward, game_over)

        
        
        ##### FILL IN HERE
        
      # Save as a mp4
      env.draw(prefix+str(e))

      # Update stats
      score = score + win-lose

      print("Win/lose count {}/{}. Average score ({})"
            .format(win, lose, score/(1+e)))
      print('Final score: '+str(score/epochs))

In [0]:
# Evaluation
testexp(agent,env,epochs_test,prefix='cnn_test_explore')
HTML(display_videos('cnn_test_explore0.mp4'))

Win/lose count 3.5/0. Average score (3.5)
Final score: 0.35
Win/lose count 30.0/3.0. Average score (15.25)
Final score: 3.05
Win/lose count 25.5/1.0. Average score (18.333333333333332)
Final score: 5.5
Win/lose count 15.0/1.0. Average score (17.25)
Final score: 6.9
Win/lose count 31.5/0. Average score (20.1)
Final score: 10.05
Win/lose count 3.5/1.0. Average score (17.166666666666668)
Final score: 10.3
Win/lose count 36.0/2.0. Average score (19.571428571428573)
Final score: 13.7
Win/lose count 14.0/0. Average score (18.875)
Final score: 15.1
Win/lose count 13.5/1.0. Average score (18.166666666666668)
Final score: 16.35
Win/lose count 19.0/2.0. Average score (18.05)
Final score: 18.05


We can see here that adding the malus board makes the average score. For temperature= 0.5 , CNN without exploration incentives reached an average score of 9.9. With exploration incentives, mainly the negative reward coming from malus_position, we reach and average score of 18.05.

***
***
__BONUS question__ Use the expert DQN from the previous question to generate some winning games. Train a model that mimicks its behavior. Compare the performances.

***